pkg/helm/repoHTTP.go

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2023 The Kubebb Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package helm
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"helm.sh/helm/v3/pkg/repo"
+
+	"sigs.k8s.io/yaml"
+)
+
+// RepoAddHTTP adds a chart repository
+// inspire by https://github.com/helm/helm/blob/dbc6d8e20fe1d58d50e6ed30f09a04a77e4c68db/cmd/helm/repo_add.go
+// some difference with `helm repo add` command
+// 1. when the same repo name is added, it will overwrite
+// 2. many options we do not need now are not supported yet.
+func RepoAddHTTP(ctx context.Context, logger logr.Logger, name, url, username, password string, httpRequestTimeout time.Duration) (err error) {
```

Comment:
There are 3 functions here, `RepoAdd` `RepoAddHTTP` and `RepoAddOCI`, and they are all the same except for a couple of different lines. Please consider streamlining this, the tests for these functions need to be updated in sync as well, also, `RepoAdd` doesn't look like it's being used anymore, so if this function is redundant, just remove it!

Reply:
The problem in this part is that there are several defer functions in the common parts and one is in an if statement. If the common part is taken out, the new function needs to return around eight variables. Would that be a little too much?

Reply:
```bash
$ diff3 repoadd repoaddhttp repoaddoci
====
1:1c
  func RepoAdd(ctx context.Context, logger logr.Logger, name, url, username, password string, httpRequestTimeout time.Duration) (err error) {
2:1c
  func RepoAddHTTP(ctx context.Context, logger logr.Logger, name, url, username, password string, httpRequestTimeout time.Duration) (err error) {
3:1c
  func RepoAddOCI(ctx context.Context, logger logr.Logger, name, url, username, password string, httpRequestTimeout time.Duration) (err error) {
====
1:53,55c
        if _, err := r.DownloadIndexFile(); err != nil && !registry.IsOCI(url) {
                return errors.Wrapf(err, "looks like %q is not a valid chart repository or cannot be reached", entry.URL)
        }
2:53,55c
        if _, err := r.DownloadIndexFile(); err != nil {
                return errors.Wrapf(err, "looks like %q is not a valid chart repository or cannot be reached", entry.URL)
        }
3:52a
```
From the three-way comparison, the three functions only have different function names and 53-55 lines. Everything else is the same.

IMO, keep `RepoAdd` and check `url` before `r.DownloadIndexFile` is enough.

Reply:
@Henry-Gao-2004 For this part, @Abirdcfly is right. We do not need to split helm operations to RepoHttp and RepoOCI here. 

1. remove `DownloadIndexFile` from `RepoAdd` as it is not required during `add repo`
2. compatible  two modes`http` and `oci` in `RepoUpdate`

Reply:
The three repo.go are now merged, and the functions are combined into one!